### PR TITLE
Migrate Azure Monitor project to our repository.

### PR DESCRIPTION
**Pull request by pmzara**
- Moving Azure Monitor source code to our repository.
- Integrate into appropriate pipelines

### CODE REVIEW

File: edge-modules/azure-monitor/src/ExampleDeployment.json [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            "settings": {
+              "image": "mcr.microsoft.com/azureiotedge-hub:1.0.10.1",
+              "createOptions": "{\"HostConfig\":{\"PortBindings\":{\"443/tcp\":[{\"HostPort\":\"443\"}],\"5671/tcp\":[{\"HostPort\":\"5671\"}],\"8883/tcp\":[{\"HostPort\":\"8883\"}]}}}"
+            }
+          }
+        },
+        "modules": {
+          "AzureMonitorForIotEdgeModule": {
+            "settings": {
+                "image": "mcr.microsoft.com/azuremonitor/containerinsights/ciprod:iot-0.1.3.3",
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/ExampleDeployment.json:35]: This image name is no longer correct. Should be `mcr.microsoft.com/azureiotedge-metrics-collector:0.1`. The version isn't real, but that's ok. We can fix it when we GA.

File: edge-modules/azure-monitor/src/Program.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            }
+
+            return Option.None<SortedDictionary<string, string>>();
+        }
+
+        // Test info is used in the longhaul, stress, and connectivity tests to provide contextual information when reporting.
+        // This info is passed from the vsts pipeline and needs to be parsed by the test modules.
+        // Includes information such as build numbers, ids, host platform, etc.
+        // Argument should be in the format key=value[,key=value]
+        public static SortedDictionary<string, string> ParseKeyValuePairs(string keyValuePairs, ILogger logger, bool shouldBeNonEmpty)
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/Program.cs:107]: This function is already implemented in `Microsoft.Azure.Devices.Edge.ModuleUtil.ModuleUtil`, so you can remove this impl and refer to that one.

File: edge-modules/azure-monitor/src/README.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
+# Azure Monitor for IoT Edge
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/README.md:1]: Change title to "Azure IoT Edge Metrics Collector".

File: edge-modules/azure-monitor/src/README.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+mem{}[http://host1:9001/metrics] mem{agg="p99"}[http://host2:9001/metrics]
+```
+
+Or, to allow the metrics `mem` and `cpu` regardless of tags or endpoint, the following could be added to `AllowedMetrics`:
+```
+mem cpu
+```
+
+
+## Contributing
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/README.md:123]: We can remove the ##Contributing, ##Building Code, and ## Tests sections.

File: edge-modules/azure-monitor/src/buildscript.bash [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/buildscript.bash:1]: Remove this script, looks like we don't need it.

File: edge-modules/azure-monitor/src/deployment.debug.template.json [hunk 6]
```diff
@@ -0,0 +1,138 @@
+{
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/deployment.debug.template.json:1]: This file doesn't seem useful...we already have ExampleDeployment.json.

File: edge-modules/azure-monitor/src/deployment.template.json [hunk 7]
```diff
@@ -0,0 +1,115 @@
+{
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/deployment.template.json:1]: This file also doesn't seem useful since we already have ExampleDeployment.json.

File: edge-modules/azure-monitor/src/module.json [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/module.json:1]: Remove this file...looks like it was part of the previous developer's personal config.

File: edge-modules/azure-monitor/src/write_vernum_print_script.sh [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/write_vernum_print_script.sh:1]: Can we remove this script? I don't think it's called from anywhere, and it references another script (`get_vernum.sh`) which doesn't seem to be here either...

**Review by damonbarry — CHANGES_REQUESTED** [review]
Looks good! I made several minor comments about naming, and about extra files and stuff we don't need. Let's do those now since they won't require re-testing.

Also, there are a few files here that already exist in the common library in our codebase (`Microsoft.Azure.Devices.Edge.Util`, which is under `edge-util/src/Microsoft.Azure.Devices.Edge.Util`). So you could just remove them and refer to the common ones:
- ISystemTime.cs
- Option.cs
- PeriodicTask.cs
- Preconditions.cs
- ShutdownHandler.cs
- SystemTime.cs
- OptionTest.cs
- PreconditionsTest.cs

But if that's going to require full re-testing we could also do it later.

File: edge-modules/azure-monitor/src/CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,9 @@
+
```
[the remark this answers is not in the document]
Comment (pmzara) [reply edge-modules/azure-monitor/src/CHANGELOG.md:1]: Removed.

File: edge-modules/azure-monitor/docker/windows/amd64/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+ARG EXE_DIR=.
+
+WORKDIR /app
+
+COPY $EXE_DIR/ ./
+
+USER ContainerAdministrator
+
+RUN REG add "HKLM\SYSTEM\CurrentControlSet\services\Dnscache" /v Start /t REG_DWORD /d 4 /f
```
[the remark this answers is not in the document]
Comment (pmzara) [reply edge-modules/azure-monitor/docker/windows/amd64/Dockerfile:12]: No This is how  we disable DNS caching in Windows docker file (Dylan sent me this code). Venkat asked me to add this.

File: edge-modules/azure-monitor/src/Microsoft.Azure.Devices.Edge.Azure.Monitor.csproj [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    <Content Include="../docker*/**/*.*" CopyToPublishDirectory="Always" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <None Include=".pipelines\install_deps.sh" />
+    <None Include=".pipelines\install_jq.sh" />
+    <None Include=".pipelines\pipeline.user.linux.yml" />
+    <None Include=".pipelines\pullandbuild.sh" />
+    <None Include=".pipelines\release.yaml" />
+    <None Include=".pipelines\write_vernum_script.sh" />
```
[the remark this answers is not in the document]
Comment (pmzara) [reply edge-modules/azure-monitor/src/Microsoft.Azure.Devices.Edge.Azure.Monitor.csproj:24]: Removed

File: edge-modules/azure-monitor/src/Microsoft.Azure.Devices.Edge.Azure.Monitor.csproj [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="[3.1.5]" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="[3.1.5]" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="[3.1.5]" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="[3.1.5]" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="[3.1.5]" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="[3.1.5]" />
+    <PackageReference Include="System.Runtime.Loader" Version="[4.3.0]" />
+
+    <!-- BouncyCastle.NetCore hasn't been updated for a long time -->
```
[the remark this answers is not in the document]
Comment (pmzara) [reply edge-modules/azure-monitor/src/Microsoft.Azure.Devices.Edge.Azure.Monitor.csproj:45]: Removed.

File: edge-modules/azure-monitor/src/ExampleDeployment.json (hunk 1 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/ExampleDeployment.json:35]: Changed.

File: edge-modules/azure-monitor/src/README.md (hunk 3 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/README.md:1]: Changed.

File: edge-modules/azure-monitor/src/buildscript.bash (hunk 5 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/buildscript.bash:1]: Removed.

File: edge-modules/azure-monitor/src/deployment.template.json (hunk 7 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/deployment.template.json:1]: Removed.

File: edge-modules/azure-monitor/src/module.json (hunk 8 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/module.json:1]: Removed.

File: edge-modules/azure-monitor/src/write_vernum_print_script.sh (hunk 9 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/write_vernum_print_script.sh:1]: Removed.

File: builds/misc/images-release.yaml [hunk 14]
```diff
@@ -17,7 +17,7 @@ jobs:
       build.registry.address: edgebuilds.azurecr.io
       build.registry.user: EdgeBuilds
       registry.address: edgerelease.azurecr.io
-      registry.user: EdgeRelease
+      registry.user: EdgeRelease
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:20]: Removed.

File: builds/misc/images-release.yaml [hunk 15]
```diff
@@ -252,6 +252,61 @@ jobs:
                 }
               ]
           signConfigType: inlineSignParams
+      - task: SFP.build-tasks.custom-build-task-1.EsrpCodeSigning@1
+        displayName: "Azure Metrics Collector Code Sign"
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:256]: Changed.

File: builds/misc/images-release.yaml [hunk 16]
```diff
@@ -312,6 +367,12 @@ jobs:
           name: "Diagnostics Module"
           project: IotedgeDiagnosticsDotnet
           version: $(version)
+      - template: templates/image-linux.yaml
+        parameters:
+          imageName: azureiotedge-metrics-collector
+          name: "Azure IoTEdge Metrics Collector"
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:373]: Changed.

File: builds/misc/images-release.yaml [hunk 17]
```diff
@@ -335,7 +396,7 @@ jobs:
     steps: 
       - script: |
             docker login $(build.registry.address) -u $(build.registry.user) -p $(build.registry.password)
-            docker login $(registry.address) -u $(registry.user) -p $(registry.password)
+            docker login $(registry.address) -u $(registry.user) -p $(registry.password)
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:399]: Removed.

File: builds/misc/images-release.yaml [hunk 18]
```diff
@@ -562,6 +623,61 @@ jobs:
                 }
               ]
           signConfigType: inlineSignParams
+      - task: SFP.build-tasks.custom-build-task-1.EsrpCodeSigning@1
+        displayName: "Azure Metrics Collector Code Sign"
+        inputs: 
+          ConnectedServiceName: "Azure IoT Edge Code Sign 2"
+          FolderPath: $(Build.BinariesDirectory)/publish/Microsoft.Azure.Devices.Edge.Azure.Monitor
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:630]: Changed.

File: builds/misc/images-release.yaml [hunk 19]
```diff
@@ -586,6 +702,12 @@ jobs:
           arguments: "-Name 'azureiotedge-diagnostics' -Project 'IotedgeDiagnosticsDotnet' -Version '$(version)' -Registry '$(registry.address)' -Namespace '$(namespace)' -Push"
           filePath: ./scripts/windows/build/Publish-DockerImage.ps1
           targetType: filePath
+      - task: PowerShell@2
+        displayName: "Build Image - Azure Metrics Collector - amd64"
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:706]: Dropped.

File: builds/misc/images-release.yaml [hunk 20]
```diff
@@ -609,3 +731,5 @@ jobs:
       displayName: 'Publish Edge Hub Manifest'
     - script: scripts/linux/buildManifest.sh -r '$(registry.address)' -u '$(registry.user)' -p '$(registry.password)' -v '$(version)' -t '$(System.DefaultWorkingDirectory)/edge-modules/SimulatedTemperatureSensor/docker/manifest.yaml.template' -n '$(namespace)' --tags '$(tags)'
       displayName: 'Publish Temperature Sensor Manifest'
+    - script: scripts/linux/buildManifest.sh -r '$(registry.address)' -u '$(registry.user)' -p '$(registry.password)' -v '$(version)' -t '$(System.DefaultWorkingDirectory)/edge-modules/azure-monitor/docker/manifest.yaml.template' -n '$(namespace)' --tags '$(tags)'
+      displayName: 'Publish Azure Metrics Collector Manifest'
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:735]: Dropped.

File: builds/misc/images-release.yaml [hunk 21]
```diff
@@ -562,6 +623,61 @@ jobs:
                 }
               ]
           signConfigType: inlineSignParams
+      - task: SFP.build-tasks.custom-build-task-1.EsrpCodeSigning@1
+        displayName: "Azure Metrics Collector Code Sign"
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images-release.yaml:627]: Dropped.

File: builds/misc/images.yaml [hunk 22]
```diff
@@ -180,6 +180,13 @@ jobs:
           imageName: azureiotedge-c2dmessage-tester
           project: CloudToDeviceMessageTester
 
+     # Azure Monitor
+      - template: templates/image-linux.yaml
+        parameters:
+          name: Azure Monitor Module
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images.yaml:186]: Changed.

File: builds/misc/images.yaml [hunk 23]
```diff
@@ -376,6 +383,13 @@ jobs:
           imageName: azureiotedge-c2dmessage-tester
           project: CloudToDeviceMessageTester
 
+     # Azure Monitor
+      - template: templates/image-windows.yaml
+        parameters:
+          name: Azure Monitor Module
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images.yaml:389]: Changed.

File: builds/misc/images.yaml [hunk 24]
```diff
@@ -406,3 +420,5 @@ jobs:
       displayName: 'Publish azureiotedge-diagnostics Manifest'
     - script: scripts/linux/buildManifest.sh -r $(registry.address) -u $(registry.user) -p $(registry.password) -v $(Build.BuildNumber) -t $(System.DefaultWorkingDirectory)/edgelet/iotedge-proxy/docker/manifest.yaml.template -n microsoft --tags "$(tags)"
       displayName: 'Publish azureiotedge-proxy Manifest'
+    - script: scripts/linux/buildManifest.sh -r $(registry.address) -u $(registry.user) -p $(registry.password) -v $(Build.BuildNumber) -t $(System.DefaultWorkingDirectory)/edge-modules/azure-monitor/docker/manifest.yaml.template -n microsoft --tags "$(tags)"
+      displayName: 'Publish Azure Metrics Collector Manifest'
```
[the remark this answers is not in the document]
Comment (pmzara) [reply builds/misc/images.yaml:424]: Changed.

File: edge-modules/azure-monitor/docker/linux/amd64/Dockerfile [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+# COPY --from=build-env /app/out ./
+
+# COPY --from=build-env /app/get_vernum.sh ./get_vernum.sh
+
+# RUN useradd -ms /bin/bash moduleuser
+# USER moduleuser
+
+# # add Application Insights instrumentation key into build
+# ENV APPINSIGHTS_INSTRUMENTATIONKEY=NWQzN2MxYTQtYWZmOS00NmFhLTllMTgtZDA1ODZhOWMwYzg2
+# ENTRYPOINT ["dotnet", "Microsoft.Azure.Devices.Edge.Azure.Monitor.dll"]
```
[the remark this answers is not in the document]
Comment (pmzara) [reply edge-modules/azure-monitor/docker/linux/amd64/Dockerfile:27]: Removed.

File: edge-modules/azure-monitor/src/README.md (hunk 4 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/README.md:123]: Removed.

File: edge-modules/azure-monitor/src/Program.cs (hunk 2 above)
Comment (damonbarry) [reply edge-modules/azure-monitor/src/Program.cs]: Ugh...I didn't realize `ModuleUtil` comes from the test tree. 🤕 Can you undo this particular change (restore the `ParseKeyValuePairs` method here, and remove the `ModuleUtils` reference from the csproj file). Sorry...

File: edge-modules/azure-monitor/src/README.md [hunk 26]
```diff
@@ -0,0 +1,126 @@
+# Azure IoT Edge Metrics Collector
+
+## About
+This repository contains souce code for Azure IoT Edge Metrics Collector Hub devices. It is currently in a private preview process.
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/README.md:4]: > This repository contains souce code for Azure IoT Edge Metrics Collector Hub devices.

This repository contains source code for the Azure IoT Edge Metrics module.

File: builds/misc/images-release.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (damonbarry) [reply builds/misc/images-release.yaml:630]: It looks like this didn't get changed...

File: edge-modules/azure-monitor/src/Program.cs (hunk 2 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/Program.cs]: Removed.

File: builds/misc/images-release.yaml [hunk 27]
```diff
@@ -36,7 +36,7 @@ jobs:
         displayName: "Edge Agent Code Sign"
         inputs: 
           ConnectedServiceName: "Azure IoT Edge Code Sign 2"
-          FolderPath: $(Build.BinariesDirectory)/publish/Microsoft.Azure.Devices.Edge.Agent.Service
+          FolderPath: $(Build.BinariesDirectory)\publish\Microsoft.Azure.Devices.Edge.Agent.Service
```
Comment (damonbarry) [inline builds/misc/images-release.yaml:39]: This task is in the Linux job, so revert back to forward slashes here. Only the Windows job will use the back slashes.

File: edge-modules/azure-monitor/src/ExampleDeployment.json [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+            "minDockerVersion": "v1.25",
+            "loggingOptions": "",
+            "registryCredentials": {}
+          }
+        },
+        "systemModules": {
+          "edgeAgent": {
+            "type": "docker",
+            "settings": {
+              "image": "mcr.microsoft.com/azureiotedge-agent:1.0.10.1",
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/ExampleDeployment.json:18]: Should by `mcr.microsoft.com/azureiotedge-agent:1.1`

File: edge-modules/azure-monitor/src/ExampleDeployment.json [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+              "createOptions": ""
+            }
+          },
+          "edgeHub": {
+            "type": "docker",
+            "status": "running",
+            "restartPolicy": "always",
+            "settings": {
+              "image": "mcr.microsoft.com/azureiotedge-hub:1.0.10.1",
```
Comment (damonbarry) [inline edge-modules/azure-monitor/src/ExampleDeployment.json:27]: Should be `mcr.microsoft.com/azureiotedge-hub:1.1`

File: edge-modules/azure-monitor/src/README.md (hunk 26 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/README.md]: Changed.

File: edge-modules/azure-monitor/src/ExampleDeployment.json (hunk 28 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/ExampleDeployment.json:18]: Done

File: edge-modules/azure-monitor/src/ExampleDeployment.json (hunk 29 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/ExampleDeployment.json:27]: Done

File: builds/misc/images-release.yaml (hunk 27 above)
Comment (pmzara) [reply builds/misc/images-release.yaml:39]: Done

**Review by damonbarry — APPROVED** [review]

**Review by damonbarry — DISMISSED** [review]

File: builds/misc/images.yaml [hunk 30]
```diff
@@ -180,6 +180,13 @@ jobs:
           imageName: azureiotedge-c2dmessage-tester
           project: CloudToDeviceMessageTester
 
+     # Azure Monitor
+      - template: templates/image-linux.yaml
+        parameters:
+          name: Metrics Collector
+          imageName: azureiotedge-metrics-collector
```
Comment (and-rewsmith) [inline builds/misc/images.yaml:187]: On line 159 there is another image with the same tag being published. We depend on this for test metrics.

File: edge-modules/azure-monitor/src/ExampleDeployment.json [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+                },
+                "ScrapeFrequencyInSecs": {
+                    "value": "300"
+                },
+                "AllowedMetrics": {
+                    "value": ""
+                },
+                "BlockedMetrics": {
+                    "value": ""
+                },
```
Comment (and-rewsmith) [inline edge-modules/azure-monitor/src/ExampleDeployment.json:63]: If this is serving as an example it might be a good idea to include the format that this env var takes.

File: edge-modules/azure-monitor/src/Program.cs [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+            LoggerUtil.Writer.LogInformation($"Starting metrics collector with the following settings:\r\n{Settings.Current}");
+            var transportSetting = new AmqpTransportSettings(TransportType.Amqp_Tcp_Only);
+            
+            ITransportSettings[] transportSettings = { transportSetting };
+            ModuleClient moduleClient = null;
+            try
+            {
+                moduleClient = await ModuleClient.CreateFromEnvironmentAsync(transportSettings);
+                moduleClient.ProductInfo = Constants.ProductInfo;
+                Option<SortedDictionary<string, string>> additionalTags = await GetAdditionalTagsFromTwin(moduleClient);
```
Comment (and-rewsmith) [inline edge-modules/azure-monitor/src/Program.cs:53]: @damonbarry  This was something I implemented in the test version of this module that I didn't expect to be included in the version visible to customers. I am concerned because the format in which we consume the additional tags from the twin is not polished. I forget what it is, but it is some unstructured string format (I think with semicolon delims?).

**Review by and-rewsmith — CHANGES_REQUESTED** [review]
I am a bit out of the loop here so I gave this a look-over. The blocking problem I see is that the build pipeline will overwrite the test-version of the metrics collector. We should choose a different name for one of these modules.

Some other concerns:
- The "additionalTags" approach of consuming metrics from the module twin was implemented by Philip and I for the tests. Therefore the format we consume the additional tags in is not ideal for customer use. I didn't expect it to be included in the customer facing monitor module.
- Seems there are a bit of TODOs. A lot of these are open ended questions where it seemed unsure of what approach to take with some concern. Might be worth giving these a once over.
-  We should make sure the build images pipelines still works, including the release pipeline.

File: builds/misc/images.yaml (hunk 30 above)
Comment (and-rewsmith) [reply builds/misc/images.yaml:187]: @pmzara I think fixing this is necessary for merge. Lmk if you need any help. Bear and I have a ton of context here.

File: edge-modules/azure-monitor/src/ExampleDeployment.json (hunk 31 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/ExampleDeployment.json:63]: Empty is the default value for this env. If the developer needs more details he/she can find them in the README.md file.

File: edge-modules/azure-monitor/src/Program.cs (hunk 32 above)
Comment (pmzara) [reply edge-modules/azure-monitor/src/Program.cs:53]: @damonbarry Should I remove or keep it?

Comment (and-rewsmith) [reply edge-modules/azure-monitor/src/Program.cs:53]: My vote is that we should limit this PRs scope to just adding the monitor module to the repo. We can make this fix in another PR.

Comment (damonbarry) [reply edge-modules/azure-monitor/src/Program.cs:53]: @pmzara I agree with Andrew's latest comment, no changes necessary here. We can address the test module later.

File: builds/misc/images.yaml (hunk 30 above)
Comment (damonbarry) [reply builds/misc/images.yaml:187]: So if I understand correctly, the Build Images pipeline will overwrite the old metrics collector (test) module with the new one (Azure Monitor). In that case, this PR _is_ the right place to address this...sorry for the back-and-forth. @and-rewsmith and @pmzara, can you decide on a docker image name for the test module, then update `builds/misc/images.yaml` _in this PR_ with the new name?

Comment (and-rewsmith) [reply builds/misc/images.yaml:187]: I think we should go with azureiotedge-test-metrics-collector

File: edge-modules/azure-monitor/src/Program.cs [hunk 33]
```diff
@@ -85,24 +84,6 @@ static async Task<int> MainAsync()
...
-        {
-            Twin twin = await moduleClient.GetTwinAsync();
-            TwinCollection desiredProperties = twin.Properties.Desired;
-            LoggerUtil.Writer.LogInformation($"Received {desiredProperties.Count} tags from module twin's desired properties that will be added to scraped metrics");
-
-            const string additionalTagsPlaceholder = "additionalTags";
-            Dictionary<string, string> deserializedTwin = JsonConvert.DeserializeObject<Dictionary<string, string>>(twin.Properties.Desired.ToJson());
-            if (deserializedTwin.ContainsKey(additionalTagsPlaceholder))
-            {
-                return Option.Some<SortedDictionary<string, string>>(ParseKeyValuePairs(deserializedTwin[additionalTagsPlaceholder], LoggerUtil.Writer, true));
```
Comment (and-rewsmith) [inline edge-modules/azure-monitor/src/Program.cs]: Unfortunately the logic for ```ParseKeyValuePairs``` is still in the module. I think we should revert the change that remove ```additionalTags``` and leave it in for now.

After the module is checked in initially we can remove this functionality in a separate PR.

**Review by and-rewsmith — COMMENTED** [review]
Almost there! @pmzara I think we need to:
- revert the change that removes the additionalTags logic
- address the comments (i.e. removing base image dockerfiles and fixing typos)
- run the build pipeline to verify nothing is broken

After that it is safe to merge

**Comment by and-rewsmith** [comment]
@vadim-kovalyov Pedro closed this PR and opened a dup that was merged. Closing this now.
